File: app/src/main/java/com/zxcx/zhizhe/ui/welcome/GuidePageActivity.java
package com.zxcx.zhizhe.ui.welcome;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.zxcx.zhizhe.R;
import com.zxcx.zhizhe.mvpBase.BaseActivity;
import com.zxcx.zhizhe.ui.MainActivity;
import com.zxcx.zhizhe.utils.Utils;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class GuidePageActivity extends BaseActivity {

    /** 引导页面的控件 */
    @BindView(R.id.viewPager)
    ViewPager mViewPager;
    /**
     * 5张引导页面的图片
     */
    private int[] mImgIds = new int[]{R.drawable.guide_1, R.drawable.guide_2,
            R.drawable.guide_3};
    /**
     * 图片资源容器
     */
    private List<ImageView> mImageViews = new ArrayList<>();

    private boolean isScrolling;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_guide_page);
        ButterKnife.bind(this);

        initImgData();

        initViewPager();
    }

    @Override
    public void initStatusBar() {
        //覆盖父类修改状态栏方法
    }

    private void initImgData() {
        for (int imgId : mImgIds) {
            ImageView imageView = new ImageView(getApplicationContext());
            imageView.setScaleType(ImageView.ScaleType.CENTER_CROP);
            imageView.setImageResource(imgId);

            mImageViews.add(imageView);
        }
    }


    /**
     * 跳转到App主界面
     */
    private void jumpToIndexActivity() {

        Intent intent = new Intent(this, MainActivity.class);
        intent.putExtra("isFirst", true);
        startActivity(intent);

        Utils.setIsFirstLaunchApp(false);

        finish();
    }


    private void initViewPager() {


        mViewPager.setAdapter(new PagerAdapter() {

            @Override
            public Object instantiateItem(ViewGroup container, int position) {
                // 给最后一页引导页添加点击事件
                if (position == mViewPager.getAdapter().getCount() - 1) {
                    View view = LayoutInflater.from(GuidePageActivity.this).inflate(R.layout.guide_5_layout,container,false);
                    ImageView guideButton = (ImageView) view.findViewById(R.id.guide_button);
                    guideButton.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            jumpToIndexActivity();
                        }
                    });
                    container.addView(view);
                    return view;
                }else {
                    container.addView(mImageViews.get(position));

                    return mImageViews.get(position);
                }

            }

            @Override
            public void destroyItem(ViewGroup container, int position,
                                    Object object) {
                container.removeView(mImageViews.get(position));
            }

            @Override
            public boolean isViewFromObject(View view, Object object) {
                return view == object;
            }

            @Override
            public int getCount() {
                return mImgIds.length;
            }
        });

        mViewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
                if(position == mImgIds.length -1 && positionOffset==0 && positionOffsetPixels == 0 && isScrolling){
                    isScrolling = false;
                    jumpToIndexActivity();
                }
            }

            @Override
            public void onPageSelected(int position) {

            }

            @Override
            public void onPageScrollStateChanged(int state) {
                if (state == 1){
                    isScrolling = true;
                }else {
                    isScrolling = false;
                }
            }
        });
    }
}

File: app/src/main/java/com/zxcx/zhizhe/ui/welcome/GuidePageActivity.kt
package com.zxcx.zhizhe.ui.welcome

import android.content.Intent
import android.os.Bundle
import android.support.v4.view.PagerAdapter
import android.support.v4.view.ViewPager
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import butterknife.ButterKnife
import com.gyf.barlibrary.ImmersionBar
import com.zxcx.zhizhe.R
import com.zxcx.zhizhe.mvpBase.BaseActivity
import com.zxcx.zhizhe.ui.MainActivity
import com.zxcx.zhizhe.utils.Utils
import kotlinx.android.synthetic.main.activity_guide_page.*
import java.util.*

/**
 * 启动引导页
 */

class GuidePageActivity : BaseActivity() {

    /**
     * 3张引导页面的图片
     */
    private val mImgIds = intArrayOf(R.drawable.guide2, R.drawable.guide3, R.drawable.guide4, R.drawable.guide1)
    /**
     * 图片资源容器
     */
    private val mImageViews = ArrayList<ImageView>()

    private var isScrolling: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_guide_page)
        ButterKnife.bind(this)

        initImgData()

        initViewPager()
    }

    override fun initStatusBar() {
        //覆盖父类修改状态栏方法
        mImmersionBar = ImmersionBar.with(this)
        mImmersionBar.keyboardEnable(true)
        mImmersionBar.init()
    }

    private fun initImgData() {
        for (imgId in mImgIds) {
            val imageView = ImageView(applicationContext)
            imageView.scaleType = ImageView.ScaleType.CENTER_CROP
            imageView.setImageResource(imgId)


            mImageViews.add(imageView)
        }
    }


    /**
     * 跳转到App主界面
     */
    private fun jumpToIndexActivity(needLogin: Boolean) {

        val intent = Intent(this, MainActivity::class.java)
        intent.putExtra("needLogin", needLogin)
        startActivity(intent)

        Utils.setIsFirstLaunchApp(false)

        finish()
    }


    private fun initViewPager() {


        view_pager.adapter = object : PagerAdapter() {

            override fun instantiateItem(container: ViewGroup, position: Int): Any {
                container.addView(mImageViews[position])
                return mImageViews[position]
            }

            override fun destroyItem(container: ViewGroup, position: Int,
                                     `object`: Any) {
                container.removeView(mImageViews[position])
            }

            override fun isViewFromObject(view: View, `object`: Any): Boolean {
                return view === `object`
            }

            override fun getCount(): Int {
                return mImgIds.size
            }
        }

        view_pager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
                if (position == mImgIds.size - 1 && positionOffset == 0f && positionOffsetPixels == 0 && isScrolling) {
                    isScrolling = false
                    jumpToIndexActivity(false)
                }

                when (position) {
                    0 -> {
                        iv1.setImageResource(R.drawable.bg_backbutton)
                        iv2.setImageResource(R.drawable.bg_backbutton4)
                        iv3.setImageResource(R.drawable.bg_backbutton4)
                        iv4.setImageResource(R.drawable.bg_backbutton4)
                    }

                    1 -> {
                        iv1.setImageResource(R.drawable.bg_backbutton4)
                        iv2.setImageResource(R.drawable.bg_backbutton)
                        iv3.setImageResource(R.drawable.bg_backbutton4)
                        iv4.setImageResource(R.drawable.bg_backbutton4)
                    }

                    2 -> {
                        iv1.setImageResource(R.drawable.bg_backbutton4)
                        iv2.setImageResource(R.drawable.bg_backbutton4)
                        iv3.setImageResource(R.drawable.bg_backbutton)
                        iv4.setImageResource(R.drawable.bg_backbutton4)
                    }

                    3 -> {
                        iv1.setImageResource(R.drawable.bg_backbutton4)
                        iv2.setImageResource(R.drawable.bg_backbutton4)
                        iv3.setImageResource(R.drawable.bg_backbutton4)
                        iv4.setImageResource(R.drawable.bg_backbutton)
                    }
                }
            }

            override fun onPageSelected(position: Int) {

            }

            override fun onPageScrollStateChanged(state: Int) {
                isScrolling = state == 1
            }
        })
    }

    override fun setListener() {
        super.setListener()
        tv_skip.setOnClickListener {
            jumpToIndexActivity(true)
        }
    }
}
